lib/edge_impulse/ei_ncs.cpp

```diff
@@ -0,0 +1,404 @@
+
+BUILD_ASSERT(CONFIG_EI_NCS_THREAD_STACK_SIZE > 0);
+
+#define DATA_BUFFER_SIZE	CONFIG_EI_NCS_DATA_BUF_SIZE
+#define THREAD_STACK_SIZE	CONFIG_EI_NCS_THREAD_STACK_SIZE
+#define THREAD_PRIORITY 	CONFIG_EI_NCS_THREAD_PRIORITY
+#define DEBUG_MODE		IS_ENABLED(CONFIG_EI_NCS_DEBUG_MODE)
+
+enum state {
+	STATE_DISABLED,
+	STATE_PROCESSING,
+	STATE_READY,
+};
+
+struct data_buffer {
+	float buf[DATA_BUFFER_SIZE];
+	float *processing_head;
+	float *append_head;
+	size_t wait_data_size;
+	struct k_spinlock lock;
+	enum state state;
+};
+
+static K_THREAD_STACK_DEFINE(thread_stack, THREAD_STACK_SIZE);
+static struct k_thread thread;
+static k_tid_t ei_thread_id;
+
+static K_SEM_DEFINE(ei_sem, 0, 1);
+
+static struct data_buffer ei_input;
+static ei_impulse_result_t ei_result;
+static ei_ncs_result_ready_cb user_cb;
+
+
+BUILD_ASSERT(DATA_BUFFER_SIZE > INPUT_WINDOW_SIZE);
+BUILD_ASSERT(INPUT_WINDOW_SIZE % INPUT_FRAME_SIZE == 0);
+
+
+static size_t buf_calc_max_processing_move(struct data_buffer *b)
+{
+	size_t res;
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	if (b->append_head >= b->processing_head) {
+		res = b->append_head - b->processing_head;
+	} else {
+		res = (buf_end - b->processing_head) +
+		      (b->append_head - buf_start);
+	}
+
+	return res;
+}
+
+static size_t buf_calc_free_space(struct data_buffer *b)
+{
+	size_t res;
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	if (b->wait_data_size > 0) {
+		res = b->wait_data_size + ARRAY_SIZE(b->buf) -
+		      INPUT_WINDOW_SIZE - 1;
+	} else {
+		if (b->processing_head > b->append_head) {
+			res = b->processing_head - b->append_head - 1;
+		} else {
+			res = (buf_end - b->append_head) +
+			      (b->processing_head - buf_start) - 1;
+		}
+	}
+
+	return res;
+}
+
+static void buf_processing_end(struct data_buffer *b)
+{
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	__ASSERT_NO_MSG(b->state == STATE_PROCESSING);
+	b->state = STATE_READY;
+
+	k_spin_unlock(&b->lock, key);
+}
+
+static int buf_cleanup(struct data_buffer *b)
+{
+	int err = 0;
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	if (b->state == STATE_PROCESSING) {
+		err = -EBUSY;
+	} else {
+		b->processing_head = b->buf;
+		b->append_head = b->buf;
+		b->wait_data_size = 0;
+		b->state = STATE_READY;
+	}
+
+	k_spin_unlock(&b->lock, key);
+
+	return err;
+}
+
+static int buf_append(struct data_buffer *b, const float *data, size_t len)
+{
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	if (buf_calc_free_space(b) < len) {
+		k_spin_unlock(&b->lock, key);
+		return -ENOMEM;
+	}
+
+	float *cur_head = b->append_head;
+	float *new_head = b->append_head + len;
+	int result = 0;
+	bool looped = false;
+
+	if (b->wait_data_size > 0) {
+		if (b->wait_data_size > len) {
+			b->wait_data_size -= len;
+		} else {
+			b->wait_data_size = 0;
+			result = PROCESS_BUF;
```

Comment:
`wait_data_size` units from data should be ignored
considering that at this stage we lost this information I doubt it is done

Reply:
Actually, `wait_data_size` informs how many floats we need before we start processing data. This data is not dropped, it's still stored in the buffer.
I think it would be better to inform in API when processing has started to make user aware about it (instead of dropping data).